### PR TITLE
MANTA-5143 Data integrity issue with Java SDK 3.4.2

### PR DESCRIPTION
**Pull request by kellymclaughlin**
This corrects a bug introduced in #562 that prevented correct decryption of some objects encrypted using AES CTR mode. The objects susceptible to this bug are ones where IV values generated during the encryption process are interpreted as negative `BigInteger` values. The calculation of applying the CTR mode offset to the initial, randomly-generated IV and the conversion back to a byte array yielded incorrect results in such cases and prevented correct decryption of the associated objects. This PR changes the method used to determine the IV value to use for decrypting the request to instead use operations directly on the byte array rather than converting the values to `BigInteger` values.

This PR also adds testing to ensure that the IV calculation when using an offset of zero results in an IV that is identical to the input IV.

### CODE REVIEW

File: java-manta-client-unshaded/src/test/java/com/joyent/manta/client/crypto/AesCtrCipherDetailsTest.java [hunk 1]
```diff
@@ -200,4 +202,36 @@ protected void canRandomlyReadPlaintextPositionFromCiphertext(final SecretKey se
...
+            Cipher cipher = cipherDetails.getCipher();
+            SecretKey secretKey = SecretKeyUtils.generate(cipherDetails);
+            cipher.init(Cipher.DECRYPT_MODE, secretKey, spec);
+            Assert.assertEquals(iv, cipher.getIV(), "Calculated initialization " +
+                    "vector does not match expected value [expected=" + Arrays.toString(iv) +
+                    " actual=" + Arrays.toString(cipher.getIV()) + "]");
+
+            // Calculate BigInteger value of IV
+            final BigInteger ivBigInt = new BigInteger(iv);
+            if (ivBigInt.compareTo(BigInteger.ZERO) == -1) {
```
[the remark this answers is not in the document]
Comment (kellymclaughlin) [reply java-manta-client-unshaded/src/test/java/com/joyent/manta/client/crypto/AesCtrCipherDetailsTest.java]: Done

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/crypto/AesCtrCipherDetails.java [hunk 2]
```diff
@@ -133,15 +132,29 @@ public AlgorithmParameterSpec getEncryptionParameterSpec(final byte[] iv, final
...
+        // Loop over the IV starting with the least significant byte to ensure
+        // our carry values are correctly calculated.
+        int carry = 0;
+        for (int ivIndex = ivLength - 1; ivIndex >= 0; ivIndex--) {
+            // Widen the byte types to int types for the arithmetic. The bit
+            // mask is used so that the addition is not performed on negative
+            // values because all Java types are signed.
+            int newIVByteAsInt = ((int)iv[ivIndex] & 0xff) + ((int)startingBlockArray[ivIndex] & 0xff) + carry;
+            byte newIVByte = (byte) newIVByteAsInt;
+            carry = newIVByteAsInt >>> 8;
```
[the remark this answers is not in the document]
Comment (kellymclaughlin) [reply java-manta-client-unshaded/src/main/java/com/joyent/manta/client/crypto/AesCtrCipherDetails.java:151]: Done

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/crypto/AesCtrCipherDetails.java [hunk 3]
```diff
@@ -133,15 +132,29 @@ public AlgorithmParameterSpec getEncryptionParameterSpec(final byte[] iv, final
...
+        byte[] startingBlockArray = ByteBuffer.allocate(ivLength).putLong(8, startingBlock).array();
+        byte[] updatedIV = new byte[ivLength];
+        // Loop over the IV starting with the least significant byte to ensure
+        // our carry values are correctly calculated.
+        int carry = 0;
+        for (int ivIndex = ivLength - 1; ivIndex >= 0; ivIndex--) {
+            // Widen the byte types to int types for the arithmetic. The bit
+            // mask is used so that the addition is not performed on negative
+            // values because all Java types are signed.
+            int newIVByteAsInt = ((int)iv[ivIndex] & 0xff) + ((int)startingBlockArray[ivIndex] & 0xff) + carry;
```
[the remark this answers is not in the document]
Comment (kellymclaughlin) [reply java-manta-client-unshaded/src/main/java/com/joyent/manta/client/crypto/AesCtrCipherDetails.java:150]: I don't see much advantage in doing that for this case. Writing it inline like this is pretty common in my experience and it's not a value that should ever change so I think having to look elsewhere to reference what a `bitmask` variable represents would be more annoying for someone reading the code.

File: java-manta-client-unshaded/src/main/java/com/joyent/manta/client/crypto/AesCtrCipherDetails.java [hunk 4]
```diff
@@ -133,15 +132,29 @@ public AlgorithmParameterSpec getEncryptionParameterSpec(final byte[] iv, final
...
+            byte newIVByte = (byte) newIVByteAsInt;
+            carry = newIVByteAsInt >>> 8;
+            updatedIV[ivIndex] = (byte)newIVByte;
```
[the remark this answers is not in the document]
Comment (kellymclaughlin) [reply java-manta-client-unshaded/src/main/java/com/joyent/manta/client/crypto/AesCtrCipherDetails.java]: Yep, good catch. Fixed.
